### PR TITLE
Run CI with two different `uv` resolution strategies: `highest` and `lowest-direct`

### PR DESCRIPTION
**Pull request by janosh**
this ensures pymatgen is compatible with both the oldest and newest dependency versions it allows in `setup.py`

also fix `typing_extensions` `ImportError` reported in https://github.com/materialsproject/pymatgen/pull/3752#issuecomment-2142954988 by @kavanase @Andrew-S-Rosen

### CODE REVIEW

File: pymatgen/core/structure.py [hunk 1]
```diff
@@ -60,6 +59,7 @@
     from ase.optimize.optimize import Optimizer
     from matgl.ext.ase import TrajectoryObserver
     from numpy.typing import ArrayLike, NDArray
+    from typing_extensions import Self
```
Comment (ml-evs) [inline pymatgen/core/structure.py:62]: So we can only type check on Python 3.10? Is there no ruff rule that can be turned on that does

```suggestion
    try:
        from typing_extensions import Self
    except ImportError:
        from typing import Self
```

or something similar?

**Comment by DanielYang59** [comment]
@shyuep Sorry for the trouble caused surrounding import of `Self`. As it seems to be causing trouble in the v2024.5.31, can I ask for yanking this release and issue a new one? Apologize again for this oversight

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -25,22 +25,18 @@ jobs:
...
       fail-fast: false
       matrix:
+        # maximize CI coverage of different platforms and python versions while minimizing the
+        # total number of jobs. We run all pytest splits with the oldest supported python
+        # version (currently 3.9) on windows (seems most likely to surface errors) and with
+        # newest version (currently 3.12) on ubuntu (to get complete coverage on unix). We
+        # ignore mac-os, which is assumed to be similar to ubuntu.
+        config:
+          - { os: windows-latest, python: "3.9", resolution: highest }
+          - { os: ubuntu-latest, python: "3.12", resolution: lowest-direct }
```
Comment (DanielYang59) [inline .github/workflows/test.yml:35]: Is this a typo? Why "Python 3.12" is the "lowest-direct" while "3.9" is the "highest"?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -25,22 +25,18 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
+        # maximize CI coverage of different platforms and python versions while minimizing the
+        # total number of jobs. We run all pytest splits with the oldest supported python
+        # version (currently 3.9) on windows (seems most likely to surface errors) and with
+        # newest version (currently 3.12) on ubuntu (to get complete coverage on unix). We
+        # ignore mac-os, which is assumed to be similar to ubuntu.
```
Comment (DanielYang59) [inline .github/workflows/test.yml:32]: I think this should be removed now that it's testing `macos-latest` now?

Comment (janosh) [reply .github/workflows/test.yml:32]: good catch!

**Pull request merged**